src/sync/sync_session.cpp

```diff
@@ -838,14 +836,34 @@ SyncSession::PublicState SyncSession::state() const
 SyncSession::ConnectionState SyncSession::connection_state() const
 {
     switch (m_connection_state) {
-        case realm::sync::Session::ConnectionState::disconnected: return ConnectionState::Disconnected;
-        case realm::sync::Session::ConnectionState::connecting: return ConnectionState::Connecting;
-        case realm::sync::Session::ConnectionState::connected: return ConnectionState::Connected;
+        case sync::Session::ConnectionState::disconnected: return ConnectionState::Disconnected;
+        case sync::Session::ConnectionState::connecting: return ConnectionState::Connecting;
+        case sync::Session::ConnectionState::connected: return ConnectionState::Connected;
         default:
             REALM_UNREACHABLE();
     }
 }
 
+void SyncSession::update_configuration(SyncConfig new_config)
+{
+    while (true) {
+        std::unique_lock<std::mutex> lock(m_state_mutex);
+        if (m_state != &State::inactive) {
```

Comment:
Why have this condition + loop? Looking at the implementation for `advance_state` it sets the new state immediately? I'm not 100% sure what happens to the `m_session` variable, but if its destroying is delayed, this isn't caught by this anyway.

Reply:
Advancing to Inactive releases the lock, so after reacquiring the lock we have to recheck that the state did not get switched away from active while we weren't holding the lock.